frame/staking/src/offchain_election.rs

```diff
@@ -0,0 +1,243 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Helpers for offchain worker election.
+
+use crate::{
+	Call, CompactAssignments, Module, NominatorIndex, OffchainAccuracy, Trait, ValidatorIndex,
+};
+use codec::Encode;
+use frame_support::debug;
+use frame_system::offchain::SubmitUnsignedTransaction;
+use sp_phragmen::{
+	build_support_map, evaluate_support, reduce, Assignment, ExtendedBalance, PhragmenResult,
+	PhragmenScore,
+};
+use sp_runtime::offchain::storage::StorageValueRef;
+use sp_runtime::PerThing;
+use sp_runtime::{RuntimeAppPublic, RuntimeDebug};
+use sp_std::{convert::TryInto, prelude::*};
+
+#[derive(RuntimeDebug)]
+pub enum OffchainElectionError {
+	/// No signing key has been found on the current node that maps to a validators. This node
+	/// should not run the offchain election code.
+	NoSigningKey,
+	/// Signing operation failed.
+	SigningFailed,
+	/// Phragmen election returned None. This means less candidate that minimum number of needed
+	/// validators were present. The chain is in trouble and not much that we can do about it.
+	ElectionFailed,
+	/// The snapshot data is not available.
+	SnapshotUnavailable,
+	/// Failed to create the compact type.
+	CompactFailed,
+}
+
+/// The type of signature data encoded with the unsigned submission
+pub(crate) type SignaturePayload<'a> = (
+	&'a [ValidatorIndex],
+	&'a CompactAssignments,
+	&'a PhragmenScore,
+	&'a u32,
+);
+
+pub(crate) const OFFCHAIN_HEAD_DB: &[u8] = b"parity/staking-election/";
+const OFFCHAIN_REPEAT: u32 = 5;
+
+pub(crate) fn set_check_offchain_execution_status<T: Trait>(
```

Comment:
Just to clarify behavior so this submit a solution for every 5 blocks number no matter in which fork the 5th block appear,
* if we have two branch competing and the + 5th block is alternatively on one branch or the other we end up submitting solution for both branch.
* if we have two branch and one is always slightly in advance from the other, we end up submitting solution only for this branch.

I don't have that much knowledge about if it is a good method, like the consensus engine could be generating fork with more than 5 blocks in advance which are generally then lost in favor of a smallest fork. (like I think Babe with this second validator stuff in case the slot is empty kind of...).

So I would have expected to submit solution for forks independently, so we would submit a solution for each 5 block for each fork. (to do so I expect we just have to change persistent offchain storage to a fork aware one)

Reply:
Yeah, doing this in a fully fork-aware way is hard and a fork-aware offchain storage is a long way away.

Maybe there is a middle ground for checking common ancestors - any two forks that share a common ancestor which started the election should allow the same submissions.

So if the election was started in block X, we will submit for forks A & B as long as X is an ancestor of both.

Reply:
> Yeah, doing this in a fully fork-aware way is hard and a fork-aware offchain storage is a long way away.

That's what I remembered too, but could `StorageValueRef::local` be used here, https://substrate.dev/rustdocs/master/sp_runtime/offchain/storage/struct.StorageValueRef.html#method.local

```rust
/// Create a new reference to a value in the fork-aware local storage.
pub fn local(key: &'a [u8]) -> Self
```

Reply:
This will probably trigger https://github.com/paritytech/substrate/blob/acb1b9b3da10c73297b139fcb26cae4928b42879/client/rpc/src/offchain/mod.rs#L53 , I am trying to revive the work I did for the unique id of child trie as offchain storage (but in a better way), but I must admit I am quite slow on this one.

Reply:
I should add something to the context here that we run the offchain worker only when `now == ElectionStatus::Open(now)`, meaning that this doesn't mean that it runs every 5 block. It can run _at most_ every 5 block. Most often the limiting condition is the election window open flag and then this function. Moreover:

In the current use case, If we reorg to a previous block, I never want to re-execute this. Hence, if the last time I ran offchain was block 20 and then I reorged to 17, I want this data in offchain storage to stay 20, so 17, 18, 19,.. will all be discarded. I think this is actually the fork-unaware aka. persistent storage, hence it is used here.

Reply:
I think rob solution is good, if two chain have forked after the election has been open then the solution should be accepted by both chain. on the other hand if they forked before the election has opened then their solution can be specific to their fork.

>  I want this data in offchain storage to stay 20, so 17, 18, 19,.. will all be discarded. I think this is actually the fork-unaware aka. persistent storage, hence it is used here.

the main point I think is that it should be better to submit solutions for all chain not the only the longest one. Because it is not necessarily the longest one that will be finalized. If there is a longer chain during the whole period of election and then a shorter chain is finalized, no solution would have been submitted on this chain.

But this is not necessarily a blocker to me, on-chain phragmen will happen otherwise.
I have no idea how the above situation is rare or not. This could be fixed later or just known to exist if this is not priority.